zc.recipe.egg_/src/zc/recipe/egg/egg.py

```diff
@@ -177,14 +185,21 @@ def working_set(
     links=(),
     index=None,
     allow_hosts=('*',),
+    cache_storage=_global_working_set_cache,
 ):
     """Helper function to build a working set.
 
     Return an instance of `pkg_resources.WorkingSet`.
 
-    Results are cached in the `_working_set_cache` dict at module level.
+    Results are cached in the mapping provided in the `cache_storage` argument.
+    This defaults to a global dict at this module. If `None` is passed then
+    caching is disabled.
```

Comment:
So, I see two different and independent issues brought by you here. 

First is: 

> module level caches are dangerous by default

It's fixable by not providing a default value for the `cache_storage` argument. I don't agree with this change because I think the benefits are greater than the danger.

Second is: 

> I don't see a reason to introduce another way to do it.

It's fixable by removing the `working_set()` module level function from the API. Probably making it a non-public method named `_working_set()` or `_build_working_set()`. 

I don't like this change because I'd rather use `working_set()` as a function than instantiate the `Egg` class just to call the method. I think the module level function is a better API.

So, if I'm right, and the new proposed API is better then we have to solve the problem you mentioned: now we have two ways of doing the same thing. We could go ahead and remove the `Egg.working_set()` method from the API. 

If we would adhere strictly to "There should be one-- and preferably only one --obvious way to do it." then we could as well go ahead and remove the `Egg.working_set()` method from the API. However it seems unwise to break backwards compatibility in a PR about performance. So I propose we violate "There should be one...".

Since:

1. I want the PR approved because of the performance gains. The API improvement is a small bonus for me.
2. I don't agree with the two changes above, but they don't impact the performance gains.

Then I'll ask kindly for @leorochael  and @jamadden decide together what changes needs to be done and I'll made them (provided they are in the scope I delimited above). 

It would be even better if you do the changes yourselves, after agreeing on what needs to be changed.